xs/src/libslic3r/PerimeterGenerator.cpp

```diff
@@ -85,44 +85,66 @@ PerimeterGenerator::process()
+
+                        // detect edge case where a curve can be split in multiple small chunks.
+                        Polygons no_thin_onion = offset(last, -(float)(ext_pwidth / 2));
+                        if (no_thin_onion.size()>0 && offsets.size() > 3 * no_thin_onion.size()) {
+                            //use a sightly smaller spacing to try to drastically improve the split
+                            Polygons next_onion_secondTry = offset2(
+                                last,
+                                -(float)(ext_pwidth / 2 + ext_min_spacing / 2.5 - 1),
+                                +(float)(ext_min_spacing / 2.5 - 1));
+                            if (abs(((int32_t)offsets.size()) - ((int32_t)no_thin_onion.size())) > 
+                                2*abs(((int32_t)next_onion_secondTry.size()) - ((int32_t)no_thin_onion.size()))) {
+                                offsets = next_onion_secondTry;
+                            }
+                        }
                         
                         // the following offset2 ensures almost nothing in @thin_walls is narrower than $min_width
                         // (actually, something larger than that still may exist due to mitering or other causes)
                         coord_t min_width = scale_(this->ext_perimeter_flow.nozzle_diameter / 3);
-                        ExPolygons expp = offset2_ex(diffpp, -min_width/2, +min_width/2);
-						
-                         // compute a bit of overlap to anchor thin walls inside the print.
-                        ExPolygons anchor = intersection_ex(to_polygons(offset_ex(expp, (float)(ext_pwidth / 2))), no_thin_zone, true);
                         
-                        // the maximum thickness of our thin wall area is equal to the minimum thickness of a single loop
-                        for (ExPolygons::const_iterator ex = expp.begin(); ex != expp.end(); ++ex) {
-                            ExPolygons bounds = _clipper_ex(ClipperLib::ctUnion, (Polygons)*ex, to_polygons(anchor), true);
-							//search our bound
+                        Polygons no_thin_zone = offset(offsets, +ext_pwidth/2, jtSquare);
+                        // medial axis requires non-overlapping geometry
+                        Polygons thin_zones = diff(last, no_thin_zone, true);
+                        //don't use offset2_ex, because we don't want to merge the zones that have been separated.
+                        Polygons expp = offset(thin_zones, (float)(-min_width / 2));
+                        //we push the bits removed and put them into what we will use as our anchor
+                        if (expp.size() > 0) {
+                            no_thin_zone = diff(last, offset(expp, (float)(min_width / 2)), true);
+                        }
+                        // compute a bit of overlap to anchor thin walls inside the print.
+                        for (Polygon &ex : expp) {
```

Comment:
Should this be an ExPolygon? I'm getting a build error on my system when trying to do a clean build of your branch.